File: SudokuSolver_Java_No_Streams/src/main/sudokusolver/javanostreams/logic/simple/HiddenTriples.java
package sudokusolver.javanostreams.logic.simple;

import sudokusolver.javanostreams.Board;
import sudokusolver.javanostreams.Cell;
import sudokusolver.javanostreams.Removals;
import sudokusolver.javanostreams.RemoveCandidates;
import sudokusolver.javanostreams.SudokuNumber;
import sudokusolver.javanostreams.UnsolvedCell;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

/*
 * https://www.sudokuwiki.org/Hidden_Candidates#HT
 *
 * If three candidates exist across three cells in a unit, then those three candidates must be placed in those three
 * cells. All other candidates can be removed from those three cells.
 */
public class HiddenTriples {
    public static List<RemoveCandidates> hiddenTriples(Board<Cell> board) {
        var removals = new Removals();
        for (var unit : board.getUnits()) {
            for (var i = 0; i < SudokuNumber.values().length - 2; i++) {
                var a = SudokuNumber.values()[i];
                for (var j = i + 1; j < SudokuNumber.values().length - 1; j++) {
                    var b = SudokuNumber.values()[j];
                    for (var k = j + 1; k < SudokuNumber.values().length; k++) {
                        var c = SudokuNumber.values()[k];
                        var cells = new ArrayList<UnsolvedCell>();
                        for (var cell : unit) {
                            if (cell instanceof UnsolvedCell unsolved &&
                                    (unsolved.candidates().contains(a) ||
                                            unsolved.candidates().contains(b) ||
                                            unsolved.candidates().contains(c))
                            ) {
                                cells.add(unsolved);
                            }
                        }
                        if (cells.size() == 3) {
                            var union = EnumSet.noneOf(SudokuNumber.class);
                            for (var cell : cells) {
                                union.addAll(cell.candidates());
                            }
                            if (union.contains(a) && union.contains(b) && union.contains(c)) {
                                for (var cell : cells) {
                                    var toRemove = EnumSet.copyOf(cell.candidates());
                                    toRemove.remove(a);
                                    toRemove.remove(b);
                                    toRemove.remove(c);
                                    removals.add(cell, toRemove);
                                }
                            }
                        }
                    }
                }
            }
        }
        return removals.toList();
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/HiddenTriples.kt
package sudokusolver.kotlin.logic

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.enumMinus
import sudokusolver.kotlin.enumUnion
import sudokusolver.kotlin.mergeToRemoveCandidates
import sudokusolver.kotlin.zipEveryTriple
import java.util.EnumSet

/*
 * http://www.sudokuwiki.org/Hidden_Candidates#HT
 *
 * If three candidates exist across three cells in a unit, then those three candidates must be placed in those three
 * cells. All other candidates can be removed from those three cells.
 *
 * For each unit
 *   For each triple of candidates
 *     Find every unsolved cell that has any of the candidates
 *     If there are three cells and each candidate is found among the three cells
 *       Remove every other candidate from the three cells
 */
fun hiddenTriples(board: Board<Cell>): List<RemoveCandidates> =
    board.units.flatMap { unit ->
        SudokuNumber.values().zipEveryTriple().mapNotNull { (a, b, c) ->
            unit.filterIsInstance<UnsolvedCell>()
                .filter { a in it.candidates || b in it.candidates || c in it.candidates }
                .takeIf { it.size == 3 }
                ?.takeIf { cells ->
                    val union = enumUnion(*cells.map { it.candidates }.toTypedArray())
                    a in union && b in union && c in union
                }
                ?.flatMap { cell ->
                    (cell.candidates enumMinus EnumSet.of(a, b, c)).map { candidate -> cell to candidate }
                }
        }.flatten()
    }.mergeToRemoveCandidates()